src/main/kotlin/org/rust/ide/refactoring/introduceConstant/ui.kt

```diff
@@ -0,0 +1,157 @@
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.block
+import java.awt.Component
+import javax.swing.DefaultListCellRenderer
+import javax.swing.JList
+
+class Highlighter(private val expr: RsExpr, private val project: Project, private val editor: Editor) : JBPopupListener {
+    private val factory = RsPsiFactory(project)
+    private var highlighter: RangeHighlighter? = null
+    private var comment: PsiElement? = null
+    private val attributes = EditorColorsManager.getInstance().globalScheme.getAttributes(EditorColors.SEARCH_RESULT_ATTRIBUTES)
+
+    fun onSelect(candidate: InsertionCandidate) {
+        dropHighlighter()
+        val markupModel: MarkupModel = editor.markupModel
+
+//        val newComment = factory.createComment("<here>")
+//        project.runWriteCommandAction {
+//            point.parent.addBefore(newComment, point.anchor)
+//        }
+//        val textRange: TextRange = newComment.textRange
+//        comment = newComment
+
+        val textRange = candidate.parent.textRange
+        highlighter = markupModel.addRangeHighlighter(
+            textRange.startOffset, textRange.endOffset, HighlighterLayer.SELECTION - 1, attributes,
+            HighlighterTargetArea.EXACT_RANGE)
+    }
+
+    override fun onClosed(event: LightweightWindowEvent) {
+        dropHighlighter()
+    }
+
+    private fun dropHighlighter() {
+        highlighter?.dispose()
+
+//        comment?.let { c ->
+//            project.runWriteCommandAction {
+//                c.delete()
+//            }
+//        }
+    }
+}
+
+fun showInsertionChooser(
+    editor: Editor,
+    expr: RsExpr,
+    callback: (InsertionCandidate) -> Unit
+) {
+    val candidates = findInsertionCandidates(expr)
+    if (isUnitTestMode) {
+        callback(MOCK!!.chooseInsertionPoint(expr, candidates))
+    } else {
+        val highlighter = Highlighter(expr, expr.project, editor)
+        JBPopupFactory.getInstance()
+            .createPopupChooserBuilder(candidates)
+            .setRenderer(object : DefaultListCellRenderer() {
+                override fun getListCellRendererComponent(list: JList<*>?,
+                                                          value: Any,
+                                                          index: Int,
+                                                          isSelected: Boolean,
+                                                          cellHasFocus: Boolean): Component {
+                    val candidate = value as InsertionCandidate
+                    val text = candidate.description()
+                    return super.getListCellRendererComponent(list, text, index, isSelected, cellHasFocus)
+                }
+            })
+            .setItemSelectedCallback { value: InsertionCandidate? ->
+                if (value == null) return@setItemSelectedCallback
+                highlighter.onSelect(value)
+            }
+            .setTitle("Insert ${expr.text} into:")
```

Comment:
Maybe "Choose scope to introduce constant" (more close to java's version)
![image](https://user-images.githubusercontent.com/3221931/81499418-3ea5d980-92d4-11ea-95d8-ddaeef8e99ac.png)